src/runtime_src/core/common/drv/cu_hls.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Xilinx HLS CU
+ *
+ * Copyright (C) 2020 Xilinx, Inc.
+ *
+ * Authors: min.ma@xilinx.com
+ */
+
+#include "xrt_cu.h"
+
+#define ECHO 0
+
+static void cu_hls_wait(void *core)
+{
+	struct xrt_cu_hls *cu_hls = core;
+
+	down_interruptible(&cu_hls->sem);
+}
+
+static void cu_hls_up(void *core)
+{
+	struct xrt_cu_hls *cu_hls = core;
+
+	up(&cu_hls->sem);
+}
+
+static int cu_hls_get_credit(void *core)
+{
+	struct xrt_cu_hls *cu_hls = core;
+
+	return (cu_hls->credits) ? cu_hls->credits-- : 0;
+}
+
+static void cu_hls_put_credit(void *core, u32 count)
+{
+	struct xrt_cu_hls *cu_hls = core;
+
+	cu_hls->credits += count;
+	if (cu_hls->credits > cu_hls->max_credits)
+		cu_hls->credits = cu_hls->max_credits;
+}
+
+static void cu_hls_configure(void *core, u32 *data, size_t sz, int type)
+{
+#if ECHO
+	return;
+#else
+	struct xrt_cu_hls *cu_hls = core;
+	u32 *base_addr = cu_hls->vaddr;
+	size_t num_reg;
+	u32 i;
+
+	num_reg = sz / sizeof(u32);
+	/* Write register map, starting at base_addr + 0x10 (byte)
+	 * This based on the fact that kernel used
+	 *	0x00 -- Control Register
+	 *	0x04 and 0x08 -- Interrupt Enable Registers
+	 *	0x0C -- Interrupt Status Register
+	 * Skip the first 4 words in user regmap.
+	 */
+	for (i = 0; i < num_reg; ++i)
+		iowrite32(data[i], base_addr + 4 + i);
+#endif
+}
+
+static void cu_hls_start(void *core)
+{
+#if ECHO
+	return;
+#else
+	struct xrt_cu_hls *cu_hls = core;
+
+	/* Bit 0 -- The CU start control bit.
+	 * Write 0 to this bit will be ignored.
+	 * Until the CU is ready to take next task, this bit will reamin 1.
+	 * Once ths CU is ready, it will clear this bit.
+	 * So, if this bit is 1, it means the CU is running.
+	 */
+	iowrite32(0x1, cu_hls->vaddr);
+#endif
+}
+
+static void cu_hls_check(void *core, struct xcu_status *status)
+{
+	struct xrt_cu_hls *cu_hls = core;
+	u32 ctrl_reg;
+	u32 done_reg = 0;
+
+#if ECHO
+	done_reg = 1;
+#else
+	/* ioread32/iowrite32 is expensive! */
+	if (cu_hls->credits == cu_hls->max_credits)
+		goto out;
+
+	/* done is indicated by AP_DONE(2) alone or by AP_DONE(2) | AP_IDLE(4)
+	 * but not by AP_IDLE itself.  Since 0x10 | (0x10 | 0x100) = 0x110
+	 * checking for 0x10 is sufficient.
+	 */
+	ctrl_reg  = ioread32(cu_hls->vaddr);
+
+	if (!(ctrl_reg & CU_AP_DONE))
+		goto out;
+
+	done_reg = 1;
+#endif
+out:
+	status->num_done = done_reg;
+	status->num_ready = done_reg;
+}
+
+static struct xcu_funcs xrt_cu_hls_funcs = {
+	.get_credit	= cu_hls_get_credit,
+	.put_credit	= cu_hls_put_credit,
+	.configure	= cu_hls_configure,
+	.start		= cu_hls_start,
+	.check		= cu_hls_check,
+	.wait		= cu_hls_wait,
+	.up		= cu_hls_up,
+};
```

Comment:
please consider using existing inter-subdevice call API. introducing a side-band interaction between two subdevice is not a good design and may cause trouble in the future.
this can be done in future pull request.

Reply:
Thanks for the comment. You are right. In my mind, these call back would be CU subdevide internal and would not exposed to other subdevice. I'd follow the same interaction in next PR.